pennylane/fourier/coefficients.py

```diff
@@ -0,0 +1,160 @@
+    if not lowpass_filter:
+        return _fourier_coefficients_no_filter(f, n_inputs, degree)
+
+    if filter_threshold is None:
+        filter_threshold = 2 * degree
+
+    # Compute the fft of the function at 2x the specified degree
+    unfiltered_coeffs = _fourier_coefficients_no_filter(f, n_inputs, filter_threshold)
+
+    # Shift the frequencies so that the 0s are at the centre
+    shifted_unfiltered_coeffs = np.fft.fftshift(unfiltered_coeffs)
+
+    # Next, slice up the array so that we get only the coefficients we care about,
+    # those between -degree and degree
+    range_slices = list(
+        range(
+            filter_threshold - degree,
+            shifted_unfiltered_coeffs.shape[0] - (filter_threshold - degree),
+        )
+    )
+
+    shifted_filtered_coeffs = shifted_unfiltered_coeffs.copy()
+
+    # Go axis by axis and take only the central components
+    for axis in range(n_inputs - 1, -1, -1):
+        shifted_filtered_coeffs = np.take(shifted_filtered_coeffs, range_slices, axis=axis)
+
+    # Shift everything back into "normal" fft ordering
+    filtered_coeffs = np.fft.ifftshift(shifted_filtered_coeffs)
+
+    # Compute the inverse FFT
+    f_discrete_filtered = np.fft.ifftn(filtered_coeffs)
+
+    # Now compute the FFT again on the filtered data
+    coeffs = np.fft.fftn(f_discrete_filtered)
+
+    return coeffs
+
+
+def _fourier_coefficients_no_filter(f, n_inputs, degree):
+    r"""Computes the first :math:`2d+1` Fourier coefficients of a :math:`2\pi` periodic
+    function, where :math:`d` is the highest desired frequency in the Fourier spectrum.
+
+    This function computes the coefficients blindly without any filtering applied, and
+    is thus used as a helper function for the true ``fourier_coefficients`` function.
+
+    Args:
+        f (callable): function that takes an array of :math:`N` scalar inputs
+        n_inputs (int): number of function inputs
+        degree (int): max frequency of Fourier coeffs to be computed. For degree :math:`d`,
+            the coefficients from frequencies :math:`-d, -d+1,...0,..., d-1, d ` will be computed.
+
+    Returns:
+        array[complex]: The Fourier coefficients of the function f up to the specified degree.
+    """
+
+    # number of integer values for the indices n_i = -degree,...,0,...,degree
+    k = 2 * degree + 1
+
+    # create generator for indices nvec = (n1, ..., nN), ranging from (-d,...,-d) to (d,...,d).
+    n_range = np.array(range(-degree, degree + 1))
+    n_ranges = [n_range] * n_inputs
+    nvecs = product(*n_ranges)
+
+    # here we will collect the discretized values of function f
+    shp = tuple([k] * n_inputs)
+    f_discrete = np.zeros(shape=shp)
+
+    for nvec in nvecs:
+        nvec = np.array(nvec)
+
+        # compute the evaluation points for frequencies nvec
+        sample_points = 2 * np.pi / k * nvec
+
+        # fill discretized function array with value of f at inpts
+        f_discrete[tuple(nvec)] = f(sample_points)
```

Comment:
I imagine this might fail in other interfaces, since `f(sample_points)` won't necessarily return a float/array.

Reply:
Resolved through addition of interface tests :+1:

Reply:
:+1: